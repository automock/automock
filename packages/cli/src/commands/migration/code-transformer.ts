import type { parse } from '@babel/parser';
import type { Node, NodePath } from '@babel/traverse';
import type traverse from '@babel/traverse';
import type generate from '@babel/generator';
import type {
  CallExpression,
  File,
  Identifier,
  ImportDeclaration,
  MemberExpression,
} from '@babel/types';
import {
  awaitExpression,
  identifier,
  isArrowFunctionExpression,
  isFunctionDeclaration,
  isFunctionExpression,
  isIdentifier,
  isMemberExpression,
} from '@babel/types';
import type { ImportManager } from './import-handler';

export class CodeTransformer {
  constructor(
    private readonly parser: typeof parse,
    private readonly traverser: typeof traverse,
    private readonly generator: typeof generate,
    private readonly importManager: ImportManager
  ) {}

  public transformCode(code: string): { code: string; modified: boolean } {
    let modified = false;

    const file: File = this.parser(code, {
      sourceType: 'module',
      plugins: ['typescript'],
    }) as File;

    this.traverser(file as Node, {
      ImportDeclaration: (path: NodePath<ImportDeclaration>) => {
        modified = this.importManager.manageImports(path, modified);
      },
      MemberExpression: (path: NodePath<MemberExpression>) => {
        // Check if the object is 'TestBed' and the method called is 'create'
        if (
          isIdentifier(path.node.object) &&
          path.node.object.name === 'TestBed' &&
          isIdentifier(path.node.property) &&
          path.node.property.name === 'create'
        ) {
          // Change 'create' to 'solitary'
          path.node.property = identifier('solitary');
          modified = true;

          // Find the topmost CallExpression in the chain ending with 'compile'
          let callExpPath = path.findParent((p) =>
            p.isCallExpression()
          ) as NodePath<CallExpression>;
          while (
            (callExpPath && !isMemberExpression(callExpPath.node.callee)) ||
            (isMemberExpression(callExpPath.node.callee) &&
              ((callExpPath.node.callee as MemberExpression).property as Identifier).name !==
                'compile')
          ) {
            callExpPath = callExpPath.findParent((p) =>
              p.isCallExpression()
            ) as NodePath<CallExpression>;
          }

          if (
            callExpPath &&
            isMemberExpression(callExpPath.node.callee) &&
            ((callExpPath.node.callee as MemberExpression).property as Identifier).name ===
              'compile'
          ) {
            const expression = awaitExpression(callExpPath.node);
            callExpPath.replaceWith(expression as Node);

            // Ensure the enclosing function is marked as async
            const functionPath = callExpPath.findParent((p) => p.isFunction());

            if (
              functionPath &&
              (isFunctionDeclaration(functionPath.node) ||
                isFunctionExpression(functionPath.node) ||
                isArrowFunctionExpression(functionPath.node))
            ) {
              functionPath.node.async = true;
            }
          }
        }
      },
    });

    const output = this.generator(
      file as Node,
      {
        retainLines: true,
        concise: false,
        compact: false,
        minified: false,
        jsescOption: {
          minimal: true,
        },
        shouldPrintComment: () => true,
      },
      code
    );

    return { code: output.code, modified };
  }
}
