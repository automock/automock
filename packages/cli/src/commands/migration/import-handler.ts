import type { NodePath } from '@babel/traverse';
import type { Identifier, ImportDeclaration, ImportSpecifier, Program } from '@babel/types';
import { importDeclaration, isImportSpecifier, stringLiteral } from '@babel/types';

export class ImportManager {
  public manageImports(path: NodePath<ImportDeclaration>, modified: boolean): boolean {
    if (
      path.node.source.value === '@automock/jest' ||
      path.node.source.value === '@automock/sinon'
    ) {
      path.node.source.value = '@suites/unit';
      modified = true;
    }

    if (path.node.source.value === '@automock/core') {
      path.node.specifiers.forEach((specifier) => {
        if (
          isImportSpecifier(specifier) &&
          (specifier.imported as Identifier).name === 'UnitReference'
        ) {
          (specifier.imported as Identifier).name = 'UnitReference';
          path.node.source.value = '@suites/unit';
          modified = true;
        }
      });
    }

    // Combine imports if necessary
    this.combineImports(path.parentPath as NodePath<Program>, '@suites/unit');
    return modified;
  }

  private combineImports(programPath: NodePath<Program>, source: string): void {
    const importDeclarations = programPath
      .get('body')
      .filter((p: NodePath) => p.isImportDeclaration() && p.node.source.value === source);

    if (importDeclarations.length > 1) {
      const combinedSpecifiers = importDeclarations.reduce((acc, declarationPath) => {
        acc.push(...declarationPath.node.specifiers);
        return acc;
      }, [] as ImportSpecifier[]);

      // Create a new import declaration with combined specifiers
      const newImportDeclaration = importDeclaration(combinedSpecifiers, stringLiteral(source));

      // Replace the first import declaration with the new one
      importDeclarations[0].replaceWith(newImportDeclaration);

      // Remove the remaining import declarations
      importDeclarations.slice(1).forEach((declarationPath) => declarationPath.remove());
    }
  }
}
