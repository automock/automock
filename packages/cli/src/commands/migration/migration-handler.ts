import type * as fs from 'fs';
import * as path from 'path';
import type { CodeTransformer } from './code-transformer';

export class MigrationHandler {
  private modifiedFiles: number = 0;

  constructor(
    private readonly fileSystem: typeof fs,
    private readonly transformer: CodeTransformer
  ) {}

  public migrate(folderPath: string): void {
    this.traverseDirectory(folderPath);

    this.fileSystem.appendFileSync(
      'migration.log',
      `Summary: ${this.modifiedFiles} files were modified.\n`
    );

    console.log(
      `Migration complete. ${this.modifiedFiles} files were modified. See migration.log for details.`
    );
  }

  private updateFile(filePath: string): void {
    const code = this.fileSystem.readFileSync(filePath, 'utf8');
    const result = this.transformer.transformCode(code);

    if (result.modified) {
      this.fileSystem.writeFileSync(filePath, result.code);
      this.fileSystem.appendFileSync('migration.log', `Modified: ${filePath}\n`);

      this.modifiedFiles++;
    }
  }

  private traverseDirectory(dir: string): void {
    this.fileSystem.readdirSync(dir).forEach((file) => {
      const fullPath = path.join(dir, file);

      if (this.fileSystem.statSync(fullPath).isDirectory()) {
        this.traverseDirectory(fullPath);
      } else if (path.extname(fullPath) === '.ts') {
        this.updateFile(fullPath);
      }
    });
  }
}
