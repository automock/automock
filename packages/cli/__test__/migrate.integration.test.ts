import { parse } from '@babel/parser';
import traverse from '@babel/traverse';
import generate from '@babel/generator';
import { CodeTransformer } from '../src/commands/migration/code-transformer';
import * as fs from 'fs';
import * as path from 'path';
import { ImportManager } from '../src/commands/migration/import-handler';

describe('Code Migrator Integration Test', () => {
  let codeTransformer: CodeTransformer;

  beforeAll(() => {
    codeTransformer = new CodeTransformer(parse, traverse, generate, new ImportManager());
  });

  test('migrate', () => {
    const code = fs.readFileSync(path.join(__dirname, './assets/spec-example.ts.txt'), 'utf8');
    const result = codeTransformer.transformCode(code);

    expect(result.modified).toBe(true);
    expect(result.code).toMatchSnapshot();
  });
});
