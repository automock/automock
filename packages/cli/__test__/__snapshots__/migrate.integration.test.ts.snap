// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Code Migrator Integration Test migrate 1`] = `
"import { TestBed,
UnitReference } from "@suites/unit";

class ClassUnderTest {}

describe('just a dummy file', () => {
  let unit: ClassUnderTest;
  let unitRef: UnitReference;

  beforeAll(async () => {
    const { unitRef: ref, unit: underTest } = await TestBed.solitary<ClassUnderTest>(ClassUnderTest).
    mock<string>('CONSTANT_VALUE').
    using('arbitrary-string').
    compile();

    unitRef = ref;
    unit = underTest;
  });

  describe('when compiling the builder and turning into testing unit', () => {});
});"
`;
